sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1418,6 +1449,51 @@ def process_bundle(self, inputs, expected_outputs, parallel_uid_counter=None):
 
     return result, split_results
 
+class ParallelBundleManager(BundleManager):
+  _uid_counter = 0
+  def process_bundle(self, inputs, expected_outputs):
+    input_value = list(inputs.values())[0]
+    if isinstance(input_value, list):
```

Comment:
I changed the iterator to return a list at a new PR, which is same as before.
Previously, we returned an iterator with one element, now it returns an iterator with N elements.
We can apply the same interface when we read either from list type or GroupingBuffer type and write to output stream.
```
for i, byte_stream in enumerate(byte_streams):
    if idx is None or i == idx:
        data_out.write(byte_stream)
        if idx is not None:
          break
```
If we introduce `partition()` function to GroupingBuffer, don't we need to check the type every time when we write to output_stream because list type doesn't have this function?
I don't have a clear idea to avoid type check here at the moment though.

Reply:
I was suggesting that we create a newclass (that could subclass list if we want, but has a partition method) to accomplish this. This way the fact that there is parallelism doesn't leak through the stack up and down (e.g the BundleManager code can be used unchanged, rather than being passed all the inputs and then a(n easy to forget) flag of which ones to ignore, and also simplifies the Buffer class in that it doesn't have an extra attribute redundantly remembering the parallelism of the context it must be used in (plus all the locking, state-tracking, sleeping etc.)

Reply:
Thanks for explanation. Your suggestion brings many advantages. I was able to get rid of handling threads and reuse BundleManager as it is by writing a new class with partition() method. It is changed at the new PR.